src/proof/lfsc/lfsc_printer.cpp

```diff
@@ -202,38 +207,93 @@ void LfscPrinter::print(std::ostream& out, const ProofNode* pn)
   out << preamble.str();
-  out << "(check" << std::endl;
-  cparen << ")";
-  // print the term let list
-  printLetList(out, cparen, lbind);
+  if (d_flatten)
+  {
+    // print term lets as definitions
+    std::stringstream cparenTmp;
+    printLetList(out, cparenTmp, lbind, true);
+  }
+  else
+  {
+    // the outer check statement for the main proof
+    out << "(check" << std::endl;
+    cparen << ")";
+    // print the term let list wrapped around the body of the final proof
+    printLetList(out, cparen, lbind, false);
+  }
 
   Trace("lfsc-print-debug") << "; print asserts" << std::endl;
   // [8] print the assertions, with letification
   // the assumption identifier mapping
   for (size_t i = 0, nasserts = iasserts.size(); i < nasserts; i++)
   {
     Node ia = iasserts[i];
-    out << "(# ";
-    LfscPrintChannelOut::printAssumeId(out, i);
-    out << " (holds ";
-    printInternal(out, ia, lbind);
-    out << ")" << std::endl;
-    cparen << ")";
+    if (d_flatten)
+    {
+      out << "(declare ";
+      LfscPrintChannelOut::printAssumeId(out, i);
+      out << " (holds ";
+      printInternal(out, ia, lbind);
+      out << "))" << std::endl;
+    }
+    else
+    {
+      out << "(# ";
+      LfscPrintChannelOut::printAssumeId(out, i);
+      out << " (holds ";
+      printInternal(out, ia, lbind);
+      out << ")" << std::endl;
+      cparen << ")";
+    }
   }
 
   Trace("lfsc-print-debug") << "; print annotation" << std::endl;
   // [9] print the annotation
-  out << "(: (holds false)" << std::endl;
-  cparen << ")";
+  if (!d_flatten)
+  {
+    out << "(: (holds false)" << std::endl;
+    cparen << ")";
+  }
 
   Trace("lfsc-print-debug") << "; print proof body" << std::endl;
   // [10] print the proof body
   Assert(pn->getRule() == PfRule::SCOPE);
   // the outermost scope can be ignored (it is the scope of the assertions,
   // which are already printed above).
   LfscPrintChannelOut lout(out);
-  printProofLetify(&lout, pnBody, lbind, pletList, pletMap, passumeMap);
 
+  if (d_flatten)
+  {
+    // print the proof letification as separate check statements, followed
+    // by the main proof.
+    for (size_t i = 0; i <= pletList.size(); i++)
+    {
+      bool isFinal = (i == pletList.size());
+      const ProofNode* p = isFinal ? pnBody : pletList[i];
+      Node res = p->getResult();
+      std::stringstream resType;
+      printInternal(resType, d_tproc.convert(res), lbind);
+      out << "(check (: (holds " << resType.str() << ")" << std::endl;
+      itp = pletMap.find(p);
+      Assert(itp != pletMap.end());
+      size_t pid = itp->second;
+      // print the letified proof
+      pletMap.erase(p);
+      printProofInternal(&lout, p, lbind, pletMap, passumeMap);
+      pletMap[p] = pid;
+      out << "))" << std::endl;
+      if (!isFinal)
+      {
+        out << "(declare ";
+        LfscPrintChannelOut::printProofId(out, pid);
+        out << " (holds " << resType.str() << "))" << std::endl;
+      }
```

Comment:
Instead of
```clojure
(check (: (holds t) <proof>))
(declare p (holds t))
```
Why not?
```clojure
(define p <proof>)
(check (: (holds t) p))
```
Does this have something to do with the way LFSC erases proofs (definition bodies)?

Reply:
The two are likely equivalent in practice. The first avoids possible rechecking of the proof since the use of `p` is simply a symbol instead of a definition, although the LFSC checker should likely do the right thing performance-wise.  I haven't verified this, though.